### PR TITLE
ComTech Barricades

### LINKED ISSUE #670: barricades

### PR DESCRIPTION
**Pull request by thymol-blue**
## About the PR

This is my first PR in SS14 and also my first time doing anything with Github! Apologies in advance if I miss anything. I am leaving this as a draft because I need some help from smarter people (see bottom).

Added:
1. Basic Barricades.
2. Upgrades for Basic Barricades.
     2.1. anti-Biohazard Upgrade.
     2.2. Blast Resistance Upgrade.
     2.3. Brute Resistance Upgrade.
     2.4. Damage Modifiers for aforementioned upgrades.
     2.5. Placeholder for radial menu is added steel/metal rods/plasteel to choose upgrade.
3. Sandbag sprite (needs entrenching tool design).
4. Damaged sprites for all of the aforementioned barricades.
5. Plasteel and Metal foldable barriers (doors!).
6. Turnstile sprites (for now, they are just fences until I figure out how to make one-way doors).
7. Crafting graphs and recipes in construction menu for aforementioned items.

## Why / Balance

- Works towards Resolving #493 
- Resolves #670 
- Moves it towards resolving #485 and #486 
- Love CM13 but can't stand BYOND after playing SS14.  Super excited for this first playtest.

## Technical details

- Used DamageVisuals component to make sprites change based on health (like cracked windows!)
- Sound effects are from fences and grills.

## Media

Damaged Barricades
image

Rotated Barricades
image

Constructing and Destroying Barricades (and blowing them up!)

https://youtu.be/gV34GDBdEDQ

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

None... I think?
**Changelog**

:cl:
- add: Added barricades, barricade upgrades, barricade doors, damaged sprites, and turnstiles. Also added ways to make them!

## Todo-list

- [X] Barricades
- [X] Barricades Upgrades
- [X] Balance Input
    - [X] Crafting Cost
    - [X] Barricade health
    - [X] Barricade DamageModifers
- [X] File organization. It is my first time coding and I worried my organization might not be great/up-to-par with other PRs.
- [X] Need to add sounds from CM13

Update: Removed some todos as I realize I should atomize my PRs more. Also removed the portable walls (kept the sprites in case someone wants to work on it) as they were extremely frustrating to work with.

### CODE REVIEW

**Comment by Tunguso4ka** [comment]
as i remember PR that added radial menu support to the SS14 was closed. So you probably should use the menu that Smug used for xeno constructions/evolutions.

**Review by QuietlyWhisper — CHANGES_REQUESTED** [review]
CM14 content and files should be seperated into their respective _CM14 folder at 
`resources/prototypes/_CM14`/etc
And
`resources/textures/_CM14`/etc
It prevents conflicts from upstream when the files are updated on upstream 👍 

Also if CM13 sounds are different let us know bc theyre easy to port

**Comment by Tunguso4ka** [comment]
*_CM14, with underscore

File: Content.Server/KillTracking/KillTrackingSystem.cs
Comment (thymol-blue) [inline Content.Server/KillTracking/KillTrackingSystem.cs:1]: I have no idea why this keeps getting removed and added. I have unstaged it every commit and yet here it is. It is adding and remove the same thing so... is it fine?

File: Resources/Audio/_CM14/Structures/metalhit.ogg
Comment (QuietlyWhisper) [inline Resources/Audio/_CM14/Structures/metalhit.ogg:1]: Needs attribution
Look at any attribution.yml for format
goes in the same folder as the sound file

**Review by QuietlyWhisper — DISMISSED** [review]

File: Content.Server/KillTracking/KillTrackingSystem.cs
Comment (QuietlyWhisper) [reply Content.Server/KillTracking/KillTrackingSystem.cs:1]: fixed it for you

File: Resources/Prototypes/_CM14/Recipes/Construction/Graphs/structures/barricade_plasteel_door.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  - node: start
+    actions:
+    - !type:DeleteEntity
+    edges:
+    - to: node_barricade_plasteel_door
+      completed:
+      - !type:SnapToGrid
+        southRotation: true
+      steps:
+      - material: Plasteel
```
Comment (QuietlyWhisper) [inline Resources/Prototypes/_CM14/Recipes/Construction/Graphs/structures/barricade_plasteel_door.yml:14]: YML linter is saying Plasteel isn't the correct entity ID, it should be SheetPlasteel
```suggestion
      - material: SheetPlasteel
```

File: Resources/Prototypes/_CM14/Recipes/Construction/Graphs/structures/barricade_plasteel_door.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        amount: 5
+        doAfter: 6
+
+  - node: node_barricade_plasteel_door
+    entity: barricade_plasteel_door
+    edges:
+    - to: start
+      completed:
+        - !type:SpawnPrototype
+          prototype: Plasteel
```
Comment (QuietlyWhisper) [inline Resources/Prototypes/_CM14/Recipes/Construction/Graphs/structures/barricade_plasteel_door.yml:24]:
```suggestion
          prototype: SheetPlasteel
```

**Review by QuietlyWhisper — CHANGES_REQUESTED** [review]

**Comment by JoeHammad1844** [comment]
can you place multiple barricades on the same tile like CM13?

**Comment by QuietlyWhisper** [comment]
> can you place multiple barricades on the same tile like CM13?

didn't think about that but it would be very easy, just make them directional window style walls
though we'd need to fix this
image

**Comment by JoeHammad1844** [comment]
image

File: Resources/Prototypes/_CM14/Recipes/Construction/Graphs/structures/barricade_plasteel_door.yml (hunk 2 above)
Comment (QuietlyWhisper) [inline Resources/Prototypes/_CM14/Recipes/Construction/Graphs/structures/barricade_plasteel_door.yml:24]:
```suggestion
          prototype: SheetPlasteel1
```

**Comment by QuietlyWhisper** [comment]
> having barricades in the right direction is important for SS13 CM for two reasons
> 
> 1. grenades can only go one way through barricades
> 2. aesthetics
> 3. xenos can burrow through walls and being unable to patch that up is cringe

yeah this seems to be something we need to fix upstream and not just for cm14

**Comment by QuietlyWhisper** [comment]
@DrSmugleaf We need this for mapping, can you look into making sure these are directional window style walls? I think this upgrade system works for now and we only need the ability to close it with an empty hand as a marine but not xeno

We can probably look into making the rest of the mechanics later

**Review by DrSmugleaf — APPROVED** [review]
In the future this will need a new system for turnstiles allowing movement through one direction only, and fixing how barricade sprites are drawn so that the bottom one always draws on top of the side ones and on top of players that are behind them, but for now this is good enough, thanks for the PR!

**Comment by thymol-blue** [comment]
Thanks for the edits, I didn't want to name it Barricades to conflict with the SS14 barricades but since it's in another folder it should be fine.

Also general rule of thumb for future commits, no underscores and CM at the front for naming convention?

**Comment by DrSmugleaf** [comment]
> Thanks for the edits, I didn't want to name it Barricades to conflict with the SS14 barricades but since it's in another folder it should be fine.
> 
> Also general rule of thumb for future commits, no underscores and CM at the front for naming convention?

It depends on what you are naming, [this should have the up to date conventions](https://docs.spacestation14.com/en/general-development/codebase-info/conventions.html)
But in general, prototype ids should be `PascalCase` (id: CMBarricadeMetal), no underscores. We still haven't made a decision on whether or not every cm-specific id should be prefixed with CM or not, but its most important when there is a conflict with an id upstream.
Construction node names should be `camelCase` (nodeMetal), no underscores.
RSI states should be `snake_case` (brute_upgrade.png), though note that this doesn't apply to DamageOverlay_4.png as the code enforces a different naming convention.
.rsi folders should be `snake_case.rsi` (brute_barricade_cracks.rsi), and folders containing multiple .rsi folders should be `PascalCase` (Barricades, TrashDrinks)

**Comment by DrSmugleaf** [comment]
For copyrights as well, any folder with sound files should have an [attributions.yml file](https://github.com/CM-14/CM-14/pull/674/commits/284971be69953ef1b87f4ed4c2c8e0d8eac57636), going to fix the one for guns in another PR now.

RSI meta.json copyright should be in the format of "Taken from cmss13 at link-to-file-with-commit-id", such as ["Taken from cmss13 at https://github.com/cmss13-devs/cmss13/blob/41109aa6eebf87892e41b61418b7b06b279d04b6/icons/obj/structures/barricades.dmi"](https://github.com/CM-14/CM-14/pull/674/commits/40c5dc0365f7160076b6710472a2bc82c33585af).
This is so that we know both which file and which commit the file came from in the future.

RSIEdit will do this for you automatically if you copy a GitHub .dmi link and paste it with CTRL+V into RSIEdit.

https://github.com/CM-14/CM-14/assets/10968691/395c3e8e-5fe9-4cdf-a0be-d7542eee7708

**Review by DrSmugleaf — APPROVED** [review]

**Pull request merged**